### ISSUE TITLE
F401 false positive on submodule imports

### ISSUE THREAD

**Issue by gcbrownthrive** [astral-sh/ruff#20120]
### Summary

It's well known that importing python modules doesn't necessarily import the submodules. It appears that F401 doesn't take into account some of the edge cases due to this behavior.

In the example below, `snowflake.connector` does NOT import `snowflake.connector.pandas_tools`, so line 2 is necessary. But `F401` assumes that line 2 is unused because it thinks the function is already accessible due to line 1.

This is reproducible with the 3 lines shown in the screenshot, and the command `ruff test.py`.

<img width="743" height="210" alt="Image" src="https://github.com/user-attachments/assets/221d7840-4d1a-4766-9beb-2317475a8733" /> 

### Version

ruff 0.12.10

**Comment by ntBre** [comment]
Submodule imports are definitely a bit touchy, see https://github.com/astral-sh/ruff/issues/4656 as one other example, but I think a better solution here is either:

```py
from snowflake.connector.pandas_tools import write_pandas

write_pandas()
```

or

```py
import snowflake.connector.pandas_tools

snowflake.connector.pandas_tools.write_pandas()
```

both of which will avoid F401, so I think the diagnostic isn't exactly wrong, although the suggestion to remove the import certainly is misleading. You can also see this in a case like this:

```py
from snowflake.connector.pandas_tools import write_pandas  # F401

snowflake.connector.pandas_tools.write_pandas()
```

without the first import. Ruff is trying to tell you that the `from`-style import is unnecessary, I think.

Here are these examples in the [playground](https://play.ruff.rs/a9e0f618-e210-45e7-afb2-ec5415430a1f). I wrapped each one in a function just to make sure they don't interfere with each other.

**Comment by gcbrownthrive** [comment]
I think I see why this is complicated, but I'd say that its suggestion is definitely wrong. If I removed the import it's telling me to, the code would not run. If I ran `ruff check --fix test.py`, the code would not run. According to the docs (https://docs.astral.sh/ruff/rules/unused-import/), "Fixes to remove unused imports are safe". But this is unsafe behavior, so either the docs need to be updated or this behavior should be fixed.

**Comment by ntBre** [comment]
Oh you're right, I didn't realize this was a safe fix. That's definitely not good.

**Comment by MichaReiser** [comment]
CC: @dylwil3 it might be interesting to see if your changes fix this.

**Comment by dylwil3** [comment]
> CC: [@dylwil3](https://github.com/dylwil3) it might be interesting to see if your changes fix this.

It does not, unfortunately. As I mention in the PR summary, the change is the least ambitious possible 😄 

I think this situation would be pretty difficult for F401 to detect since it relies on very subtle behavior of the Python import system. For example, notice that

```python
from a.b import foo

a.b.foo()
```

fails with a name error because the symbol `a` is not bound. The code snippet in the OP is pretty misleading because it could equally well be written in any of the following ways and still run:

```python
import snowflake.connector
from snowflake.connector.pandas_tools import something_completely_different_from_that_module

snowflake.connector.pandas_tools.write_pandas()
```

```python
import snowflake.connector
from snowflake.connector import pandas_tools

snowflake.connector.pandas_tools.write_pandas()
```

```python
import snowflake.connector
from snowflake.connector import pandas_tools as foo

snowflake.connector.pandas_tools.write_pandas()
```

I don't even really know where that behavior is documented / whether it's part of the _specification_ or if it's an artifact of the implementation of from imports - which caches submodule imports as valid members of the top-level module, even if they are aliased. I don't think we model this behavior at all? I don't have an immediately helpful suggestion for how to proceed... but I'll try to think about it.

**Comment by danparizher** [comment]
> I don't even really know where that behavior is documented / whether it's part of the _specification_ or if it's an artifact of the implementation of from imports - which caches submodule imports as valid members of the top-level module, even if they are aliased. I don't think we model this behavior at all? I don't have an immediately helpful suggestion for how to proceed... but I'll try to think about it.

I did some digging and found that it’s documented (and specified) that importing a submodule binds it as an attribute of its parent package. The import statement reference notes that for forms like `import package.submodule`, only `package` is bound in the local namespace, and `submodule` is accessed as an attribute of `package`—which implies the binding on the parent object.

So, the behavior described isn’t merely an implementation artifact; it’s part of the language’s specified import semantics.

https://docs.python.org/3/tutorial/modules.html#packages
https://docs.python.org/3/library/importlib.html#importlib.import_module

When it comes to implementing a fix, my idea of a conservative approach could be to detect the pattern where we have both `import package` and `from package.submodule import ...` alongside qualified usage like `package.submodule.*`. In these cases, instead of suggesting removal of the submodule import, we could either suppress the F401 autofix entirely or suggest a safe rewrite like `import package.submodule` to maintain the qualified access pattern. This would keep the "Fixes to remove unused imports are safe" guarantee accurate while avoiding the complex task of modeling Python's import-time attribute binding behavior.

**Comment by danielhollas** [comment]
I found a really strange case of (what I would consider) F401 false positive.

This code work

```python
import urllib.parse
import urllib.request

def path_to_url(path: str) -> str:
    return urllib.parse.urljoin("file://", urllib.request.pathname2url(path))
```

but this reports a F401 violation

```python
import urllib.parse


def path_to_url(path: str) -> str:
    import urllib.request
    return urllib.parse.urljoin("file://", urllib.request.pathname2url(path))
```

```console
❯ ruff check urls.py 
F401 [*] `urllib.parse` imported but unused
 --> urls.py:1:8
  |
1 | import urllib.parse
  |        ^^^^^^^^^^^^
  |
help: Remove unused import: `urllib.parse`

```

Now, admittedly, `urllib.parse` does exist when one imports `urllib.requests`, but that is a import side effect that the code should not rely on. 

Not sure if this should be a separate issue
